internal/credentials/vault/gcp.go

```diff
@@ -0,0 +1,189 @@
+
+	var err error
+	gcpProject := l.authObj.Spec.GCP.ProjectId
+	if gcpProject == "" {
+		gcpProject, err = metadata.ProjectID()
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch GCP project from compute metadata: %w", err)
+		}
+	}
+	gkeLocation := l.authObj.Spec.GCP.Region
+	if gkeLocation == "" {
+		gkeLocation, err = metadata.InstanceAttributeValue("cluster-location")
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch GKE cluster location from instance metadata: %w", err)
+		}
+	}
+	gkeName := l.authObj.Spec.GCP.ClusterName
+	if gkeName == "" {
+		gkeName, err = metadata.InstanceAttributeValue("cluster-name")
+		if err != nil {
+			return nil, fmt.Errorf("failed to fetch GKE cluster name from instance metadata: %w", err)
+		}
+	}
+
+	// Retrieve the workload identity k8s service account
+	key := ctrlclient.ObjectKey{
+		Namespace: l.providerNamespace,
+		Name:      l.authObj.Spec.GCP.WorkloadIdentityServiceAccount,
+	}
+	sa, err := helpers.GetServiceAccount(ctx, client, key)
+	if err != nil {
+		logger.Error(err, "Failed to get service account")
+		return nil, err
+	}
+
+	// Create and exchange a k8s token for a Google ID token (signed jwt)
+	config := GCPTokenExchangeConfig{
+		KSA:            sa,
+		GkeClusterName: gkeName,
+		GcpProject:     gcpProject,
+		Region:         gkeLocation,
+		VaultRole:      l.authObj.Spec.GCP.Role,
+	}
+	signedJwt, err := GCPTokenExchange(ctx, config, client)
+	if err != nil {
+		return nil, fmt.Errorf("failed GCP token exchange: %w", err)
+	}
+
+	// Pass back the signed jwt as loginData
+	loginData := map[string]any{
+		"role": l.authObj.Spec.GCP.Role,
+		"jwt":  signedJwt,
+	}
+	return loginData, nil
+}
+
+type GCPTokenExchangeConfig struct {
+	KSA            *corev1.ServiceAccount
+	GkeClusterName string
+	GcpProject     string
+	Region         string
+	VaultRole      string
+}
+
+// GCPTokenExchange creates and exchanges a k8s service account token for a
+// federated access token from Google's STS API, and uses the federated access
+// token to get an ID token (signed jwt) from Google's IAM API, which can then
+// be used to auth to Vault
+func GCPTokenExchange(ctx context.Context, config GCPTokenExchangeConfig, client ctrlclient.Client) (string, error) {
+	// Read the GCP service account (GSA) from the k8s service account (KSA)
+	// annotation
+	gsa := ""
+	if v, ok := config.KSA.Annotations[GCPAnnotationServiceAccount]; ok {
+		gsa = v
+	} else {
+		return "", fmt.Errorf("workload identity service account %q is missing annotation %q", config.KSA.Name, GCPAnnotationServiceAccount)
+	}
+
+	workloadIdentityPool := fmt.Sprintf("%s.svc.id.goog", config.GcpProject)
+	k8sTokenRequest, err := helpers.RequestSAToken(ctx, client, config.KSA, 600, []string{workloadIdentityPool})
+	if err != nil {
+		return "", fmt.Errorf("failed to get service account token: %w", err)
+	}
+
+	identityProvider := fmt.Sprintf(
+		"https://container.googleapis.com/v1/projects/%s/locations/%s/clusters/%s",
```

Comment:
I don't think it's crucial as GKE will cover almost everyone, but do we know how much more work it would be to also support running from GCP VMs? (I'm assuming this URL format means we only support GKE currently)

Reply:
Yeah this only supports GKE's workload identity. We may want to support setting a static credential file or using the underlying compute credentials in GKE someday, but workload identity seemed like the best option to add first until there's sufficient demand for the others, since this is all running inside kubernetes.

For running on GCP VMs, would that just be vanilla kubernetes on GCP compute instances?

Reply:
> For running on GCP VMs, would that just be vanilla kubernetes on GCP compute instances?

That's what I was thinking of, with some kind of more generic GCE workload identity, but I think I was imagining a feature that doesn't exist, and maybe getting mixed up with [federated workload identity](https://cloud.google.com/iam/docs/workload-identity-federation-with-kubernetes#terraform) which is more about authenticating to GCP from outside GCP.